### PR TITLE
CB to LB: Add Lesson Group Description and Big Questions to Script DSL

### PR DESCRIPTION
**Pull request by dmcavoy**
Adds the ability for levelbuilders to add a description and set of big questions for a lesson group. These are both features available on CurriculumBuilder which we want to move over.

The levelbuilder adds the description and question using the following DSL syntax (A levelbuilder can have as many questions as they want).

```
lesson_group 'lg-1', display_name: 'Lesson Group 1'
lesson_group_description 'This is the description of the lesson group.'
lesson_group_question 'What is question 1?'
lesson_group_question 'What is question 2?'
```

The description and questions are saved in the scripts.en.yml file so they can be translated. 

## Links

- [spec](https://docs.google.com/document/d/13ZQuoNKMR0sgJsIVjHRSUVgfRnpIadfHwG8-DONoyxE/edit#bookmark=id.qj6vv1ekf9ul)
- [jira](https://codedotorg.atlassian.net/browse/PLAT-248)

## Testing story

Added tests going from DSL and creating the objects as well as starting with objects and creating the DSL.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/dsl/script_dsl.rb [hunk 1]
```diff
@@ -81,11 +81,22 @@ def lesson_group(key, properties = {})
...
         key: key,
         display_name: properties[:display_name],
+        big_questions: [],
         lessons: []
       }.compact
     end
   end
 
+  def lesson_group_description(description)
+    current_lesson_group = @lesson_groups.length - 1
```
Comment (uponthesun) [inline dashboard/app/dsl/script_dsl.rb]: optional: could use `@lesson_groups.last` here instead

File: dashboard/app/dsl/script_dsl.rb [hunk 2]
```diff
@@ -275,7 +286,15 @@ def i18n_hash
     i18n_lesson_group_strings = {}
     @lesson_groups.each do |lesson_group|
       if lesson_group[:key]
-        i18n_lesson_group_strings[lesson_group[:key]] = {'display_name' => lesson_group[:display_name]}
+        i18n_lesson_group_strings[lesson_group[:key]] = {
+          'display_name' => lesson_group[:display_name]
+        }
+        if lesson_group[:description]
+          i18n_lesson_group_strings[lesson_group[:key]][:description] = lesson_group[:description]
```
Comment (uponthesun) [inline dashboard/app/dsl/script_dsl.rb]: (Optional)

For these three things (display_name, description, big_questions) it seems like we're just directly copying it over into the new hash in `i18n_lesson_group_strings`, with the same keys? Could we simplify this a bit by looping through those three keys and doing the same thing for each?

Or maybe using `select`? https://apidock.com/ruby/v2_6_3/Hash/select

File: dashboard/test/models/script_test.rb [hunk 3]
```diff
@@ -2177,6 +2177,24 @@ class SummarizeVisibleAfterScriptTests < ActiveSupport::TestCase
     assert_equal 'required', script.lessons[0].lesson_group.key
   end
 
+  test 'can add description and big questions for lesson group' do
+    l = create :level
+    dsl = <<-SCRIPT
+      lesson_group 'lg-1', display_name: 'Lesson Group'
```
Comment (uponthesun) [inline dashboard/test/models/script_test.rb:2183]: Maybe add a second lesson group to this test case to test that descriptions and big questions are correctly associated with their lesson group?

**Review by uponthesun — COMMENTED** [review]
Looks good to me overall, I mostly have minor comments.

It does feel like between script_dsl and unit tests, there's a lot of code we need to change when we add new properties, and it's mostly boilerplate. Do we have a sense of how many similar changes we'll need to make for CB to LB? Could it be worth investing in refactoring things to make adding new properties easier and more declarative? @dmcavoy @davidsbailey 

Also, can you add/update test cases on the changes to the `i18n_hash` method?

File: dashboard/app/dsl/script_dsl.rb (hunk 2 above)
Comment (Hamms) [reply dashboard/app/dsl/script_dsl.rb]: :heavy_plus_sign:

**Comment by dmcavoy** [comment]
> Looks good to me overall, I mostly have minor comments.
> 
> It does feel like between script_dsl and unit tests, there's a lot of code we need to change when we add new properties, and it's mostly boilerplate. Do we have a sense of how many similar changes we'll need to make for CB to LB? Could it be worth investing in refactoring things to make adding new properties easier and more declarative? @dmcavoy @davidsbailey

I think we are not planning to use DSL for anything besides the stuff on the script edit page (script, lesson groups, key and name of lesson) so I don't think we will be adding more properties this way beyond the additions to script. @davidsbailey do you agree?

> Also, can you add/update test cases on the changes to the `i18n_hash` method?

Updated an existing test with description and big questions.

File: dashboard/app/models/lesson_group.rb [hunk 4]
```diff
@@ -36,6 +36,8 @@ class LessonGroup < ApplicationRecord
 
   serialized_attrs %w(
     display_name
+    description
+    big_questions
```
Comment (davidsbailey) [inline dashboard/app/models/lesson_group.rb:40]: happy we have a great place to store this info now! 🎉

File: dashboard/test/dsl/script_dsl_test.rb [hunk 5]
```diff
@@ -593,7 +593,11 @@ def setup
   test 'Script DSL for lesson with lesson group' do
     input_dsl = <<~DSL
       lesson_group 'required', display_name: 'Overview'
+      lesson_group_description 'This is a description'
+      lesson_group_question 'Question 1'
+      lesson_group_question 'Question 2'
```
Comment (davidsbailey) [inline dashboard/test/dsl/script_dsl_test.rb:598]: sad we're using DSL to store this. but agree this is the best path forward.

when we get to editing lessons, we shouldn't be adding anything new as DSL.

**Review by davidsbailey — APPROVED** [review]

**Comment by davidsbailey** [comment]
> > Could it be worth investing in refactoring things to make adding new properties easier and more declarative? @dmcavoy @davidsbailey
> 
> I think we are not planning to use DSL for anything besides the stuff on the script edit page (script, lesson groups, key and name of lesson) so I don't think we will be adding more properties this way beyond the additions to script. @davidsbailey do you agree?

Agreed, we don't want to add any more DSL properties after these ones, and then we want to move off of DSL as an editing interface and storage format, so in general this is not an area we want to make a big investment in. the seed logic refactoring was a bit of an exception because it was necessary in order to get us onto the new data model which included lesson group (though one could argue that we could have deferred that too). So I think we would take a quick fix to speed up this process if there is one, but I am skeptical that such a fix would be quick.

> I don't think we will be adding more properties this way beyond the additions to script.

@dmcavoy are there specific properties you have in mind to add to script? I am not seeing anything else to add to script DSL in the [Moving Unit & Chapter functionality from CB -> LB](https://docs.google.com/document/d/13ZQuoNKMR0sgJsIVjHRSUVgfRnpIadfHwG8-DONoyxE/edit#heading=h.jl29htiuqe0x) spec, but I might be misunderstanding what you're saying.

**Comment by dmcavoy** [comment]
> @dmcavoy are there specific properties you have in mind to add to script? I am not seeing anything else to add to script DSL in the [Moving Unit & Chapter functionality from CB -> LB](https://docs.google.com/document/d/13ZQuoNKMR0sgJsIVjHRSUVgfRnpIadfHwG8-DONoyxE/edit#heading=h.jl29htiuqe0x) spec, but I might be misunderstanding what you're saying.

I didn't have anything specific in mind so there might not being anything. It was more of a general statement of where things could get added. @davidsbailey

File: dashboard/app/dsl/script_dsl.rb [hunk 6]
```diff
@@ -286,14 +284,9 @@ def i18n_hash
...
-        i18n_lesson_group_strings[lesson_group[:key]] = {
-          'display_name' => lesson_group[:display_name]
-        }
-        if lesson_group[:description]
-          i18n_lesson_group_strings[lesson_group[:key]][:description] = lesson_group[:description]
-        end
-        unless lesson_group[:big_questions].nil_or_empty?
-          i18n_lesson_group_strings[lesson_group[:key]][:big_questions] = lesson_group[:big_questions]
+        i18n_lesson_group_strings[lesson_group[:key]] = {}
+        lesson_group.select {|k, _| [:display_name, :description, :big_questions].include?(k)}.each do |k, v|
```
Comment (uponthesun) [inline dashboard/app/dsl/script_dsl.rb]: Could this be simplified even more by directly assigning the entire result of `select`? Something like:

```
i18n_lesson_group_strings[lesson_group[:key]] = lesson_group.select ...
```

Also, could the `nil_or_empty` check go into the condition inside the select block?

**Review by uponthesun — APPROVED** [review]

File: dashboard/app/dsl/script_dsl.rb (hunk 6 above)
Comment (dmcavoy) [reply dashboard/app/dsl/script_dsl.rb]: I moved the nil or empty check but I'm not sure how to make the other change work. I'm going to merge it this way and can always follow up.

Comment (uponthesun) [reply dashboard/app/dsl/script_dsl.rb]: It's completely fine as is, but if you were curious, I was thinking something like this:

```
        keys = [:display_name, :description, :big_questions]
        strings = lesson_group.select {|k, v| keys.include?(k) && !v.nil_or_empty?}.stringify_keys
        i18n_lesson_group_strings[lesson_group[:key]] = strings
```

(Lines 2 and 3 on separate lines purely for readability / line length reasons)

YMMV on whether that's better, though. I generally prefer doing things more "functionally" if it's easy to do. The nice thing here is if I just read `strings = lesson_group.select ...`, even without seeing the rest of the block, I know that `strings` is just going to be a subset of the `lesson_group` hash. Whereas anything can happen in a for loop.

Comment (dmcavoy) [reply dashboard/app/dsl/script_dsl.rb]: TIL `stringify_keys` thats definitely a helpful thing. Thanks for sharing this winter!

**Pull request merged**